### ISSUE TITLE
Proposal to simlify module imports

### ISSUE THREAD

**Issue by Gozala** [LuxLang/lux#39]
@LuxLang I'm writing down a specific proposal with code snippets that I'm  happy to post here most up to date version, but can we discuss collaborate on draft on google docs or dropbox paper instead ? In my experience it provides a lot better interface for proposing edits and inline feedback.

**Comment by LuxLang** [comment]
@Gozala Sure.

**Comment by Gozala** [comment]
Here is the link to the current draft: https://docs.google.com/document/d/12ilq2c6lUI0FsFkYXuVvEx8uKBpf0vduUDuvit_mw1k/edit?usp=sharing

Anyone can comment or edit in suggestion mode, meaning that edits will be submitted as proposed change.

**Comment by Gozala** [comment]
I think first draft is ready!

**Comment by LuxLang** [comment]
I'll be annotating it with comments related to the current design, to add some context.

That could help find some middle point that we can all agree on.

**Comment by Gozala** [comment]
> I'll be annotating it with comments related to the current design, to add some context.
> 
> That could help find some middle point that we can all agree on.

Thanks! And by all means I expected it to require many refinements until it could reach something that can be adopted.

**Comment by Gozala** [comment]
Moving discussion thread from the document here as google docs is better and providing edit suggestions but not great at lengthy discussions.

> @mccraigmccraig:
> is this true ? if import it is just a convenience that implies there is a global namespace, and a global namespace is the root of jar-hell (since it prevents solutions like node's provision of packages as values, permitting different versions of a given module to be resident concurrently)
> @gozala:
> I would expect compiler being able to namespaced usage `name/space/thing` and treat as if there was a `(import name/space name/space)` all along. So nothing prevents here from node style, although there might be some other issues that I'm not aware of.
> @mccraigmccraig:
> ah, yeah, that works
> @eduardoejp:
> I'm not sure that is such a wise idea.

> If you think about it, making that assumption implies that the compiler has scanned and analysed all the .lux files it has in its scope (be they in your codebase, or in imported libraries), and can correlate every namespace/module you use to wherever it shows up in any file available.

> This is highly problematic.

> Explicit imports solve the problem (at least when modules have a 1-to-1 mapping with .lux files), because the compiler doesn't have to analyse every file available anywhere, but is instead pointed to the specific files containing the definitions.

> This eliminates any possible ambiguity, and the need for global search.

I'll respond to the comment in the document here instead of trying to fit it inside tho google doc thing:

> I'm not sure that is such a wise idea.

> If you think about it, making that assumption implies that the compiler has scanned and analysed all the .lux files it has in its scope (be they in your codebase, or in imported libraries), and can correlate every namespace/module you use to wherever it shows up in any file available.

> This is highly problematic.

> Explicit imports solve the problem (at least when modules have a 1-to-1 mapping with .lux files), because the compiler doesn't have to analyse every file available anywhere, but is instead pointed to the specific files containing the definitions.

> This eliminates any possible ambiguity, and the need for global search.

**Comment by Gozala** [comment]
> If you think about it, making that assumption implies that the compiler has scanned and analysed all the .lux files it has in its scope (be they in your codebase, or in imported libraries), and can correlate every namespace/module you use to wherever it shows up in any file available.

I am not sure I'm totally following, so I'll try to offer my view and maybe pointing out when it breaks down would help me understand the issue.

My expectation that was that compiler will start with a entry point file and follow the dependency tree from there. My assumption is that reader will mark all instances like `module.fn` or `path/to/module.fn` as namespaced identifiers and presumably build a table of externals. At the analyzer pass compiler will resolve each external according to either imports or if import isn't found by resoling a module corresponding to that namespace.

In other words I don't understand why compiler would have to analyze anything other that what's being imported explicitly or implicitly by referring to it via path.

**Comment by Gozala** [comment]
That not to say that it does not complicates things, it does as imports aren't just declared but rather they can be scattered across the forms with-in the module and it well could be that complexity it would add the compiler is not worth the benefit of saving few keystrokes. 

@eduardoejp if you think that is a case I can modify document to drop this and make import mandatory. That being said it would mean that would introduce ambiguity a following instance:

```clj
(lux.import
  lux
  other
  core [log!] lux
  [program:] lux/cli)
```

Which maybe an argument in support of option suggested by @mccraigmccraig, that is:
```clj
(lux.import
  lux
  other
  (core [log!] lux)
  ([program:] lux/cli))
```
Please let me know what do you thing & I'll adjust proposal accordingly.

**Comment by LuxLang** [comment]
@Gozala 
The issue is that you write symbols like "my/module.my-function", but then the compiler has to figure out where that definition is located.

I mean, if definition "my/module.my-function" exists, then somewhere there must be a module called "my/module".
The question is: _where?_

If imports are not specified, then how will the compiler find such a module?

Now, you could say: _but, just look for a file called "my/module.lux", and use it as your module._

OK, that's nice.

But, if modules are now logical units, and there isn't a 1-to-1 correspondence between modules and .lux files, then that means I cannot just look for a file called "my/module.lux" (such a file may not even exist, or the definitions in the module "my/module" may be scattered across multiple files).

That is why I said that not having explicit imports and 1-to-1 mapping between modules and .lux files would force the compiler to do a complete scan of all available .lux files.

Otherwise, how will the compiler know where a particular definition is located?

**Comment by LuxLang** [comment]
OK, time to talk a bit about the current design of the Lux module system.

So, basically, the current system just maps the _logical concept_ of module, to the _physical reality_ of a `.lux` file.
If a module is named `foo`, it is assumed that _somewhere accessible_ there exists a file named `foo.lux`.
If a module is named `foo/bar/baz`, it is assumed that _somewhere accessible_ there exists a file named `foo/bar/baz.lux`.

The slashes imply directories and files (or other directories) contained in them, as in a file-system.
With that said, you don't have unrestricted access to anywhere in your file-system, as all the files/modules are relative to a kind of _fictional root_ established by the compiler.
What that means is that if `foo.lux` is inside your `/source` directory, while `bar.lux` is inside your `/test` directory and `bar.lux` is inside some library package imported by the build tool, as far as the compiler is concerned, they are all on the same level, and thus are children of the _fictional root_.

This is to avoid any complexity from expressing everything in terms of an actual file-system.

Now, this kind of way of structuring modules and mapping them 1-to-1 to files is nothing special, considering that most modern languages do just that.

---

Now, the interesting thing about separating the logical notion of a module from the physical notion of a file, is that it could open the possibility to new kinds of uses to modules.
Currently, a module just provides a way to access and refer to values (specified as **definitions** in the source code contained in particular `.lux` files).

A question could be: _what if modules could be used for arbitrary code organization, regardless of what file it came from?_

And that does sound like an interesting prospect, which is why I engage in this debate.

But an issue I have with that (which I have talked about to some degree before) is my worry that by divorcing the logical concept of a module from its physical representation, it might become harder for the compiler to figure out where any of the definitions you're referring to are located.

For example: _the possibility of referring to definitions without explicitly importing their containing file has been previously mentioned._
This is highly problematic, as it implies that either the compiler possesses supernatural divination powers to find out where the code for that definition is located, or that it would perform some kind of global scan to find things for you.
( If there is another way to find the code without explicit imports, feel free to mention it. )

There is also another problem that even if you force people to use explicit imports (and just treat it as importing a file, while still keeping the concept of modules semantically separate), you still run into the problem that if a module's definitions are scattered across multiple files, just knowing that a file **F** contains definitions of a module **M** is not enough (it could be the case that **M**'s definitions are scattered across multiple files, so you might also have to import files **F2**, **F3**, ..., **Fn** before you finally get what you want).

There are also problems such as:
1. If you can define new definitions for an already existing module somewhere else, that could easily create a monkey-patching culture, with people tinkering with already existing modules in libraries/frameworks for the sake of convenience. _I am personally skeptical of the benefits of this practice._
2. Having the file hierarchy be disconnected from the module hierarchy might introduce some measure of confusion, as searching for the modules you want won't be straightforward (e.g. you're searching for the module `lux/data/number/complex`, but it turns out there is no file `lux/data/number/complex.lux`, so you look up the file `lux/data/number.lux`, but it's not there either. Turns out, somebody defined it inside `lux/math.lux`, but that wasn't so obvious).

I also have the preoccupation that part of the motivation behind separating modules from files is purely aesthetic and is a matter of taste, instead of actual functionality or objective advantages.
I say this because an example usage that seems like a motivating factor for this proposal is being able to open structures and make the resulting definitions look like something they came out of a module, instead of a structure.
The proposal seems to go as far as _defining structures as modules_ in the first place, and letting the compiler take care of handling their usage both as structures and modules.

To be honest, I see this purely as an aesthetic choice, with no real impact in terms of functionality, and actual value provided by the language.
For this reason I am somewhat skeptical as to the merits of this example, since it seems based purely on personal preference, and not any actual objective benefits.

Specially considering that it would require re-engineering parts of the language and compiler, to what amounts to no actual practical gain.

Granted, I may be speaking out of personal bias for the current way of doing things.
But I would personally demand better examples of things this divorce between _logical modules_ and _physical files_ enables, before considering it as an actual good idea to implement.

Fortunately, better examples have been provided (for example, being able to define sub-modules that are all about testing, or other ancillary concerns in your code).

Like: _being able to define the tests for module `foo` in the same file, but under a dependent module `foo/test`_.

This sounds interesting, but the question of how you handle the ambiguity introduced by disconnecting _logical modules_ from _physical files_, which I talked about previously, is still a factor for concern.

---

There also seems to be some measure of anxiety involved in this proposal, in the form of the desire to make everything in Lux be a value (one complaint raised in the Gitter channel being that modules are not first-class entities, and that this somehow diminishes the elegance of the language).

I have to say that I'm not particularly concerned with this.
The way I see it, something being a value does not make it special just because it's a value.
It is only what _value semantics_ enable that makes it worthwhile to _value-ify_ things.

Currently, I don't see the big benefit in turning Lux modules into first-class entities.

And I don't say that to be dismissive.

Currently, from what I understand of the proposal, it seems it's trying to blur the lines between modules and structures (to me, it doesn't even seem like blurring, but like downright erasing those lines).
First-class modules behave like structures, and structures get to be defined just like modules.

But, is this in any way related to modules and their purpose on the language?

What do first-class modules actually gain us?
* If you say that they now behave like structures: _then they have gained us nothing._
Structures already behave like structures. Why do we need other things to behave like structures?
* If you say that they are now first-class values and you can pass them around: _then they have gained us nothing._
First-class _things_ are not valuable because you can pass them around, but because they enable new forms of composition.

First-class functions are valuable because they _enable_ function composition.
First-class types are valuable because they _enable_ type-based computation.

What do first-class modules _enable_?

Perhaps creating new modules at run-time...

Well, we can already create new structures at run-time, so that use-case is already covered.

The problem with modules is that they are means for logically organizing code, but what does it mean to create them at run-time?
Does it mean that the program can expand itself at run-time, or maybe re-define its own definitions dynamically.

That sounds interesting, but it requires either an interpreter or a VM made to handle that kind of stuff.
Perhaps, once the Lux compiler has been integrated into the standard library, dynamic run-time compilation could _enable_ this kind of stuff.

But adding that capacity at the language level would imply **pretty deep and far-reaching consequences** for Lux, affecting its performance, portability, security, and many other concerns.
It is not a light decision to make.

---

Personally, I worry that this proposal, of first-class modules, might be trying to solve some problems the wrong way.

Personally, I am skeptical that it will turn out as a good idea in the end (at least in its current form).

---

**Note: If I misunderstood any of the points being made, or ignored anything during my critique, please feel free to correct me.**

**Comment by lostbean** [comment]
I agree with @eduardoejp in keeping import explicit and 1:1 mapping with the file structure. Allowing multiple modules per file would only be cool for single file/very small projects.

But I really like @Gozala's proposition of simplifying import. And, in my opinion, the following syntax is very simple and clean:

```
(lux.import
  lux
  other
  (core [log!] lux)
  ([program:] lux/cli))
```

I'm also very much in favor of replacing `;` by `.` and being explicit about `lux` module (e.g. `;something` would be `lux.something`)

**Comment by Gozala** [comment]
My apologies for not getting back to this thread sooner, it's just I had lots going on in personal life. Unfortunately I won't be able to address everything right now so I'll have to keep it short & address all of the misunderstandings in the proposal document itself.



> @Gozala
> The issue is that you write symbols like "my/module.my-function", but then the compiler has to figure out where that definition is located.
> 
> I mean, if definition "my/module.my-function" exists, then somewhere there must be a module called "my/module".
> The question is: where?
> 
> If imports are not specified, then how will the compiler find such a module?
> 
> Now, you could say: but, just look for a file called "my/module.lux", and use it as your module.
> 
> OK, that's nice.
> 
> But, if modules are now logical units, and there isn't a 1-to-1 correspondence between modules and .lux files, then that means I cannot just look for a file called "my/module.lux" (such a file may not even exist, 

I think I may not have described intended behavior well enough. I would expect that compiler will consider doing following:



1. Look for `module` sub-module under  `my.lux` file.
2. Look into `module/my.lux` file

Ideas is to start with one and fallback to other but which option takes precedence is up to debate IMO each has it's own pros and cons.


> or the definitions in the module "my/module" may be scattered across multiple files).
> 

I definitely did not want to propose to allow definitions of the same module across the different files. But I see how it could have being misunderstood. Idea is that file path is always considered, submodules can add sub/entry to it and split it's definitions across that file but file path is always there to make discovery possible as described above. So say if I had following code under the path `mylib/math/util.lux`:

```clj
(lux.module:
 (lux.import
  [assert] lux/test
  [=] lux/data/number/Eq<Nat>)


 (def: square
   (-> Nat Nat)
   (lambda [n] (n.* n n)))

 (lux.module: test
   (assert (= (square +1) +1))
   (assert (= (square +2) +2)))


 (lambda factorial' [acc n]
   (if (n.= +0 n)
     acc
     (factorial' (n.* n acc) (n.dec n))))

 (def: factorial
   (-> Nat Nat)
   (factorial' +1))

 (lux.module: test
   (assert (= (factorial +0) +0))
   (assert (= (factorial +5) +120)))
```

I'd be able to import test via path `mylib/math/util/test` or refer to function `factorial` under `mylib/math/util.factorial`.

There is a possibility that  you'd have `mylib/math.lux` with a `util` submodule but in case of  `mylib/math/util/test` it will be clear it's not submodule given that there's no sub-sub-modules. In case of `mylib/math/util.factorial` it would have to assume one and fallback to other, but I think that might actually be beneficial for runtime specific things.

> That is why I said that not having explicit imports and 1-to-1 mapping between modules and .lux files would force the compiler to do a complete scan of all available .lux files.
> 
> Otherwise, how will the compiler know where a particular definition is located?

I hope my comments above clarified it a bit
